fairseq/sequence_generator.py

```diff
@@ -212,9 +213,9 @@ def _generate(
         bsz, src_len = src_tokens.size()[:2]
         beam_size = self.beam_size
 
-        if constraints is not None and not self.search.supports_constraints:
+        if (constraints is not None or negative_constraints is not None) and not self.search.supports_constraints:
             raise NotImplementedError(
-                "Target-side constraints were provided, but search method doesn't support them"
+                "Target-side constraints or negative constraints were provided, but search method doesn't support them"
             )
 
         # Initialize constraints, when active
```

Comment:
You can modify the API for `init_constraints` to pass `negative_constraints` as well

Reply:
Thank you for your detailed suggestions. What I originally wanted to do was not change the existing code as much as possible, but just add something. But it also seems to change some APIs, and I try to move the changes to the interior of each part. I will create a new push soon.